File: src/providers/tinfoil.ts
import axios from "axios";
import zlib from "zlib";
import { promisify } from "util";
import fs from "fs/promises";
import path from "path";
import yaml from "js-yaml";
import type { AttestationReport, ServiceProvider } from "../types";

const gunzip = promisify(zlib.gunzip);

export class TinfoilProvider implements ServiceProvider {
	name = "tinfoil";
	private configPath = path.join(__dirname, "../config/tinfoil_config.yml");
	private cache: Record<string, string> | null = null;

	private async getModelMap(): Promise<Record<string, string>> {
		if (this.cache) return this.cache;

		try {
			const fileContent = await fs.readFile(this.configPath, "utf-8");
			// biome-ignore lint/suspicious/noExplicitAny: YAML config is dynamic
			const config = yaml.load(fileContent) as any;

			const map: Record<string, string> = {};
			// Config structure:
			// models:
			//   model-name:
			//     enclaves: [host...]

			if (config.models) {
				for (const [key, value] of Object.entries(config.models)) {
					// biome-ignore lint/suspicious/noExplicitAny: YAML config is dynamic
					const modelData = value as any;
					if (
						modelData.enclaves &&
						Array.isArray(modelData.enclaves) &&
						modelData.enclaves.length > 0
					) {
						map[key] = modelData.enclaves[0];
					}
				}
			}

			this.cache = map;
			return map;
		} catch (err) {
			console.error("Failed to load Tinfoil config:", err);
			// Fallback to empty or throw? Throwing is safer as it means broken setup.
			throw new Error("Failed to load Tinfoil configuration");
		}
	}

	async fetchReport(modelId: string): Promise<AttestationReport> {
		const map = await this.getModelMap();
		const host = map[modelId];

		if (!host) {
			throw new Error(`Unknown Tinfoil model: ${modelId}`);
		}

		// Allow direct hostname usage if the modelId looks like a host (legacy support)
		const actualHost = modelId.includes(".") ? modelId : host;

		const url = `https://${actualHost}/.well-known/tinfoil-attestation`;
		console.log(`[Tinfoil] Fetching from ${url}`);

		const response = await axios.get(url);
		const data = response.data;
		const expectedPrefix = "https://tinfoil.sh/predicate/tdx-guest/";

		if (!data.format || !data.format.startsWith(expectedPrefix)) {
			throw new Error(
				`Unsupported Tinfoil attestation format: ${data.format || "missing"}`,
			);
		}

		if (!data.body) {
			throw new Error("Tinfoil response missing body");
		}

		const buffer = Buffer.from(data.body, "base64");
		const decoded = await gunzip(buffer);

		// Decoded is the raw Intel TDX Quote (binary)
		const quoteHex = decoded.toString("hex");

		return {
			intel_quote: quoteHex,
			nvidia_payload: undefined,
			raw: data,
		};
	}

	async listModels(): Promise<string[]> {
		const map = await this.getModelMap();
		return Object.keys(map);
	}
}

File: confidential_verifier/providers/tinfoil.py
import requests
import base64
import gzip
import yaml
import os
from typing import List, Dict, Any, Optional
from .base import ServiceProvider
from ..types import AttestationReport
from ..verifiers.tinfoil import TinfoilVerifier
from ..verifiers import Verifier


class TinfoilProvider(ServiceProvider):
    """
    Provider for Tinfoil attestation.

    Tinfoil runs enclaves on both Intel TDX and AMD SEV-SNP hardware.
    The provider tries the specific enclave endpoint first, then falls back
    to the central router.
    """

    ROUTER_URL = "https://inference.tinfoil.sh"

    def __init__(self, config_path: Optional[str] = None):
        if not config_path:
            config_path = os.path.join(
                os.path.dirname(__file__), "../../config/tinfoil_config.yml"
            )
            if not os.path.exists(config_path):
                config_path = "config/tinfoil_config.yml"

        self.config_path = config_path
        self._cache = None

    def get_verifier(self) -> Verifier:
        return TinfoilVerifier()

    def _get_model_config(self) -> Dict[str, Dict[str, Any]]:
        """Load model configuration from YAML file."""
        if self._cache is not None:
            return self._cache

        try:
            with open(self.config_path, "r") as f:
                config = yaml.safe_load(f)

            self._cache = config.get("models", {})
            return self._cache
        except Exception as e:
            print(f"Failed to load Tinfoil config from {self.config_path}: {e}")
            return {}

    def _fetch_attestation(self, url: str) -> Dict[str, Any]:
        """Fetch and parse attestation from a URL."""
        response = requests.get(url, timeout=10)
        response.raise_for_status()
        return response.json()

    def fetch_report(self, model_id: str) -> AttestationReport:
        """
        Fetch attestation report from Tinfoil.

        Tries the specific enclave endpoint first, then falls back to the router.
        Supports both Intel TDX and AMD SEV-SNP formats.
        """
        model_config = self._get_model_config()
        repo = None
        enclave_url = None

        # Try to get enclave endpoint from config
        if model_id in model_config:
            repo = model_config[model_id].get("repo")
            enclaves = model_config[model_id].get("enclaves", [])
            if enclaves:
                enclave_url = f"https://{enclaves[0]}/.well-known/tinfoil-attestation"

        data = None

        # Try enclave endpoint first
        if enclave_url:
            try:
                print(f"[Tinfoil] Trying enclave: {enclave_url}")
                data = self._fetch_attestation(enclave_url)
                print(f"[Tinfoil] Successfully fetched from enclave")
            except Exception as e:
                print(f"[Tinfoil] Enclave failed ({e}), falling back to router")

        # Fall back to router
        if data is None:
            url = f"{self.ROUTER_URL}/.well-known/tinfoil-attestation"
            print(f"[Tinfoil] Fetching from router: {url}")
            data = self._fetch_attestation(url)

        fmt = data.get("format", "")
        body = data.get("body")

        if not body:
            raise Exception("Tinfoil response missing body")

        # Decompress the quote
        compressed_quote = base64.b64decode(body)
        quote_bytes = gzip.decompress(compressed_quote)

        # Detect attestation format
        if "sev-snp" in fmt:
            quote_type = "sev-snp"
            print(f"[Tinfoil] Detected AMD SEV-SNP attestation")
        elif "tdx" in fmt:
            quote_type = "tdx"
            print(f"[Tinfoil] Detected Intel TDX attestation")
        else:
            quote_type = "unknown"
            print(f"[Tinfoil] Unknown attestation format: {fmt}")

        data["repo"] = repo
        data["model_id"] = model_id
        data["quote_type"] = quote_type

        return AttestationReport(
            provider="tinfoil",
            model_id=model_id,
            intel_quote=quote_bytes.hex(),  # Field name kept for compatibility
            nvidia_payload=None,
            raw=data,
        )

    def list_models(self) -> List[str]:
        """List models from config file."""
        return list(self._get_model_config().keys())
